File: leetcode/189-rotate-array/RotateArray.java
public class RotateArray {
    private void reverse(int[] nums, int i, int j) {
        while (i < j) {
            nums[i] ^= nums[j];
            nums[j] ^= nums[i];
            nums[i] ^= nums[j];

            i++;
            j--;
        }
    }

    public void rotate(int[] nums, int k) {
        int n = nums.length;
        int index = k % n;

        reverse(nums, 0, n - 1);
        reverse(nums, 0, index - 1);
        reverse(nums, index, n - 1);
    }
}

File: leetcode/189-rotate-array/rotate-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
const rotate = function (nums, k) {
  k = k % nums.length;

  reverse(nums, 0, nums.length - 1);
  reverse(nums, 0, k - 1);
  reverse(nums, k, nums.length - 1);
};

const reverse = function (nums, start, end) {
  let left = start;
  let right = end;

  while (left < right) {
    [nums[left], nums[right]] = [nums[right], nums[left]];
    left++;
    right--;
  }
};
